### ISSUE TITLE
Difficulty getting the vat API's to function

### ISSUE THREAD

**Issue by Comtass** [hmrc/vat-api#191]
While I managed to get through the hello-world examples fine and am confident about refreshing OAuth tokens etc when it comes to getting the actual API's to work I am struggling. To start with Obligations example, it connects fine, the authorisation is working - I know this as if I omit the token refresh it doesn't work - but it keeps returning	{"statusCode":400,"message":"INVALID_DATE_FROM"}	
The request body is:-
{"from":"2017-01-25",
   "to":"2018-05-31",
   "status":"F"}
Until I can get past this obstacle then I can't progress

**Comment by MarkAKelly** [comment]
Hi @Comtass,

Can you show me what parameters you are using on the request please?

**Comment by Comtass** [comment]
Hi Mark - I think this is what you are looking for :- 

Host: test-api.service.hmrc.gov.uk
Authorization: Bearer 7e8099385bb2944dae303a91d7f6352d
Accept: application/vnd.hmrc.1.0+json
Content-Type: application/json
Content-Length: 52

Is that correct?

Regards,

Gordon

**Comment by Comtass** [comment]
the request is:-

"GET", "/organisations/vat/" & VatRegNO & "/obligations", sbRequestBody, sbResponseBody 
The sbRequestBody Json is listed in the first message I posted

**Comment by Comtass** [comment]
Is that OK MarkAKelly  ??

**Comment by MarkAKelly** [comment]
I'm a bit unsure about how your calling this endpoint, as the GET request doesn't accept a body - I don't know if this is something in the software you're using or not.

The parameters 'from' and 'to' should be appended to the request url and are mandatory, the 'status' parameter is optional and should also be appended to the url.  This would make the request url something like this ->

'/organisations/vat/:vrn/obligations?from=2017-01-25&to=2018-05-31&status=F'

Additionally, this request might cause issues, as the request range is over 365 days.  Try again with the correct url pattern and let me know what you see.

**Comment by Comtass** [comment]
Thanks for your input - it is much appreciated.
I changes the program to append the details to the URL but the error returned now is CLIENT_OR_AGENT_NOT_AUTHORISED. 
A would appreciate your input here as hopefully I will be able to fix this

**Comment by MarkAKelly** [comment]
So this now looks like the bearer token for the user you have set up has expired.   Either use the authorisation process to create a refresh token, or create a new user with a fresh bearer token and VRN and give it a go.

**Comment by Comtass** [comment]
Well I created a new user and got a new VRN. 
It I don't refresh the token I get error 401 - Invalid Credentials, and If I do I get 403 - client or agent not authorised. I seem to be making some progress but I guess I must still be doing something wrong.
Any further ideas would be much appreciated.

**Comment by Comtass** [comment]
I have spent quite a bit of time on this and seem unable to make any progress, Can you make ant suggestions as to what I can try now?

**Comment by MarkAKelly** [comment]
It still seems like an authentication/registration issue.

All I can suggest is you follow the authentication process again and let us know what you're seeing when you're setting up your user.

**Comment by Comtass** [comment]
Hi.

I have been checking things out again and have loaded everything into PostMan so you can clearly see what is happening, and the screen is as follows:-

[cid:image001.png@01D41AB3.F7E07EA0]


I have gone through my application in the sandbox and checked that the ClientId and ClientSecret are correct in the OAuth2 procedures. The token is correct, as if not of if it times out I get Invalid Credentials rather than the error above.

It may be that I need to do something to my application in  the sandbox – I’m not sure.

Can you help me on this one please, or do I have to raise another support request?

Best regards,

Gordon

**Comment by Comtass** [comment]
Sorry - I missed your last message.
When I select to start a new OAuth2 process it opens a browser and directs me to a page headed Authority to interact with HMRC on your behalf. I then select continue and instead of sign in I click on Don't have test user credentials then I select Organisation then Create. It then gives me a userId, a password and a new VRN. The next time instead of selecting I don't have credentials I sign in with the new username & password and click on Grand Authority and I get a JSon script back in the application something like  👎 {
  "access_token": "a03e16e154b3bb205bc5f9b83bfca21",
  "refresh_token": "e7a1b4d7e882c2721804aa9fc672e10",
  "expires_in": 14400,
  "scope": "read:vat",
  "token_type": "bearer",
  "expires_on": "1531501306"
}

The full Get request is:-
test-api.service.hmrc.gov.uk/organisations/vat/666145586/obligations?from=2018-01-01&to=2018-05-31&status=F
Hoping that helps

Best regards,

Gordon

**Comment by MarkAKelly** [comment]
Right I think I can see your problem.  There are issues with creating test users using the frontend interface - we've been advising people to create their test users using the API route instead.

We did release some notes about this in our documentation, but unfortunately we don't own all of the API documentation around authentication, so it might not always be so clear.

Try creating a user with the [API](https://developer.service.hmrc.gov.uk/api-documentation/docs/api/service/api-platform-test-user/1.0) instead, and let us know how you get on.

**Comment by Comtass** [comment]
Thanks for this info Mark.
I added a section to my app to create a test user using your API and it seemed to work successfully, returning the following :- 
{
  "userId": "055615183246",
  "password": "5kKmskyismjx",
  "userFullName": "Kay Russel",
  "emailAddress": "kay.russel@example.com",
  "organisationDetails": {
    "name": "Company I3QIKT",
    "address": {
      "line1": "6 Finchley Road",
      "line2": "Dover",
      "postcode": "TS21 1PA"
    }
  },
  "saUtr": "2000013969",
  "nino": "OC900883A",
  "mtdItId": "XZIT00000477107",
  "empRef": "244/MHFBWRIKC1",
  "ctUtr": "8000072112",
  "vrn": "666869695",
  "lisaManagerReferenceNumber": "Z363750",
  "secureElectronicTransferReferenceNumber": "376060300996",
  "pensionSchemeAdministratorIdentifier": "a0181424",
  "eoriNumber": "GB4156246740"
}
 So far so good.

I then tried again to GET the obligations in my app and it kept coming back with an error. Rather than try to explain how my app works I then reverted to Postman so that I could see if the problem was with my app, but the problem persists. The following is the Postman screen which will net you see exactly what I am doing, and what is being returned:-

![image](https://user-images.githubusercontent.com/39489338/42748798-5e47d8ee-88d9-11e8-8806-184886bb72b9.png)

I can't see the image here so I have sent it as a direct email reply too.
The error I am getting is :- 
    "code": "INTERNAL_SERVER_ERROR",
    "message": "An internal server error occurred"
}
Best regards,

Gordon
